pennylane/devices/qubit/simulate.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simulate a quantum script."""
+
+# pylint: disable=protected-access
+import pennylane as qml
+from pennylane.wires import Wires
+from pennylane.measurements import StateMeasurement, MeasurementProcess
+from pennylane.typing import TensorLike
+
+from .initialize_state import create_initial_state
+from .apply_operation import apply_operation
+
+
+def measure_state_diagonalizing_gates(
+    measurementprocess: StateMeasurement, state: TensorLike
+) -> TensorLike:
+    """Apply a measurement to state when the measurement process has an observable with diagonalizing gates.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (TensorLike): state to apply the measurement to
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    total_indices = len(state.shape)
+    wires = Wires(range(total_indices))
+
+    for op in measurementprocess.obs.diagonalizing_gates():
+        state = apply_operation(op, state)
+
+    return measurementprocess.process_state(qml.math.flatten(state), wires)
+
+
+def measure(measurementprocess: MeasurementProcess, state: TensorLike) -> TensorLike:
+    """Apply a measurement process to a state.
+
+    Args:
+        measurementprocess (MeasurementProcess): measurement process to apply to the state
+        state (TensorLike): the state to measure
+
+    Returns:
+        Tensorlike: the result of the measurement
+    """
+    if isinstance(measurementprocess, StateMeasurement):
+        if measurementprocess.obs is None:
+            # no need to apply diagonalizing gates
+            total_indices = len(state.shape)
+            wires = Wires(range(total_indices))
+            return measurementprocess.process_state(qml.math.flatten(state), wires)
+
+        if measurementprocess.obs.has_diagonalizing_gates:
+            return measure_state_diagonalizing_gates(measurementprocess, state)
+
+    raise NotImplementedError
+
+
+def simulate(circuit: qml.tape.QuantumScript) -> tuple:
```

Comment:
Will we need a `shots` argument when simulating sample-based measurements?

Reply:
I was thinking of adding those in when we add shot support to this function.